packages/server-wallet/src/wallet/wallet.ts

```diff
@@ -22,16 +26,27 @@ export class Wallet {
    */
   public static async create(
     engine: Engine,
-    messageService: MessageServiceInterface,
+    messageServiceFactory: MessageServiceFactory,
```

Comment:
This is not a major point, but I wonder if the following would be a bit easier to reason about:
1. The `create` method expects a `messageService` object (not a factory).
2. The `messageService` exposes a `onNewMessage` setter. The wallet uses the setter to define the code that the `messageService` should execute when a new message arrives.

Reply:
That's an interesting thought! I think that would be a lot clearer than providing a `MessageServiceFactory`.

I think that does change a few things though:
- The `wallet` wouldn't fully own the `messageService`.  Probably not a huge deal but it would have some implications. (Should the wallet destroy the message service when it's done, etc...)
- The `messageService` could receive a message before `onNewMessage` is set. Again probably not a huge deal but we may want to think about what we want to happen in that case.
- Minor but we might want to consider what happens when `onNewMessage` is set multiple times.

Another similar option to what you're suggesting is making the `messageService` an event emitter that emits an event when it receives a message.

I think my preference would be to leave this as-is for now and revisit it outside the scope of this PR? I think you raise a really valid point that accepting a factory is less straightforward to reason about though.